Component initial denisty

onufer opened the pull request:
## Description

fixed an Issue (#819) where components initialized through the constructor would have different density than those created through blueprints. 

<!-- Mandatory: Please include a summary of the change and link to any related GitHub Issues.-->

---

## Checklist

<!--
    You (the pull requester) should put an `x` in the boxes below you have completed.
    If you're unsure about any of them, don't hesitate to ask. We're here to help!
    Learn what a "good PR" looks like here: https://terrapower.github.io/armi/developer/tooling.html#good-pull-requests
-->

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

<!-- Check the code quality -->

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

<!-- Check the project-level cruft -->

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

onufer commented on armi/reactor/reactors.py:
```diff
@@ -215,7 +215,6 @@ def __init__(self, name):
         self._circularRingPitch = 1.0
         self._automaticVariableMesh = False
         self._minMeshSizeRatio = 0.15
-        self._inputHeightsConsideredHot = True
```
unused

onufer commented:
this will resolve #819

albeanth reviewed (commented):
Here's an initial review for `component.py` and `test_components.py`. 
- There is a great deal of value in these new tests, they are great additions!
- In an effort to improve clarity, I would suggest splitting these tests off into their own test class and into individual tests. 
- I think doing so will not only help the review, but also make the tests later on (e.g. when switching between temps) clearer. Anything that is needed for other assertions (e.g., mass1) can be stored in the setUp method. 

```python 
class TestAxialExpansion(unittest.TestCase):
    """demonstrate conservation during radial and axial expansion"""
    def setUp(self):
        hotHeight = 1.0
        tWarm = 50
        coldOuterDiameter = 1.0
        circle1 = Circle("circle", "HT9", 20, tWarm, coldOuterDiameter)
        tHot = 500
        circle2 = Circle("circle", "HT9", 20, tHot, coldOuterDiameter)

    def test_massDensityProportional(self):
        """mass density is proportional to Fe number density and derived from all the number densities and atomic masses"""
        self.assertAlmostEqual(
            circle1.p.numberDensities["FE"] / circle2.p.numberDensities["FE"],
            circle1.getMassDensity() / circle2.getMassDensity(),
        )

    def test_massColdGreaterThanMassHot(self):
        """docstring"""
        mass1 = circle1.getMassDensity() * circle1.getArea() * hotHeight
        mass2 = circle2.getMassDensity() * circle2.getArea() * hotHeight
        self.assertGreater(mass1, mass2)
    
    etc...
```

opotowsky reviewed (approved):
I'd love a PR description (I'll edit to include the issue), but I think this work looks good, and solves a very important issue!

Note that the slight reduction in code coverage is because this branch is 2 commits behind main and is not indicative of lack of testing coverage in this PR.

opotowsky reviewed (changes requested):
@onufer Can you please update this bug fix in the release notes?

onufer commented on armi/reactor/components/component.py:
```diff
@@ -348,10 +348,12 @@ def applyMaterialMassFracsToNumberDensities(self):
         self.p.numberDensities = densityTools.getNDensFromMasses(
             density, self.material.p.massFrac
         )
+        self.applyHotHeightDensityReduction()
```
good point, updated.

onufer commented:
The test is meant to be a tutorial for how to expand something. it will be a lot of duplicated code to show each step individually and be much less clear to read. The idea is that it describes a tutorial how to do this type of thing in a script and what you can expect to be conserved.

As a result of your comment, i did split it into its own test class, and 1 test for height taken as hot, and 1 for height taken as cold. I hope that helps with addressing your concerns. Let me know if its more clear. Its quite a bit fewer lines than the original testing.

jakehader commented on armi/reactor/converters/axialExpansionChanger.py:
```diff
@@ -175,9 +192,10 @@ def axiallyExpandAssembly(self, thermal: bool = False):
                 b.p.zbottom = self.linked.linkedBlocks[b][0].p.ztop
-            # if not in the dummy block, get expansion factor, do alignment, and modify block
-            if ib < (numOfBlocks - 1):
+            isDummyBlock = ib == (numOfBlocks - 1)
```
Can this be based on flags?

onufer replied:
scope creep.... they are treated as flags in other places, but this is a potential separate issue.

there are warnings if they are not dummy
![image](https://user-images.githubusercontent.com/50242044/185217711-2e8f6e1f-5351-4e65-9e50-4a1b8b38601b.png)

albeanth replied:
@jakehader - I don't think it should be, actually. If we were to change it to be something like 

```python
if not b.hasFlags(Flags.DUMMY):
    # do the expansion/contraction on the block
``` 

then we'd be expanding/contracting the top-most block of an assembly if it is not have a dummy  block. This could be problematic for downstream plugins as some require the assemblies to all be the same height. 

By leaving it as is, we ensure that the top-most block, dummy or not, is chopped to preserve assembly height.

jakehader replied:
Got it! Thank you for the clarification. I'm good then, but note I haven't done a complete review of the work or tests so I am leaving that to @onufer or other reviewers unless I'm needed.

albeanth commented on armi/reactor/converters/tests/test_axialExpansionChanger.py:
```diff
@@ -765,6 +766,14 @@ def test_coldAssemblyHeight(self):
+                        bStd.getComponent(Flags.FUEL).material, custom.Custom
+                    ):
+                        # custom materials don't expand
+                        self.assertGreater(bExp.getMass("U235"), bStd.getMass("U235"))
```
Just a note to my future self. 

Since the heights in the blueprints are the same, the mass of `inputHeightsConsideredHot: False` (bExp) will be higher than `inputHeightsConsideredHot: True` (bStd). The former is cooler and therefore has a higher density than its hot counterpart - since the two cases have the same volume, the former has a higher mass. 

When bExp is expanded at BOL, mass is conserved. So post expansion, the mass is still higher than bStd.

albeanth reviewed (approved):
Coverage failure looks to me to be coverage spray. 

I think this is ready to go.
